src/Microsoft.NetCore.Analyzers/Core/Tasks/UseValueTasksCorrectly.cs

```diff
@@ -0,0 +1,547 @@
+            RuleId,
+            s_localizableTitle,
+            new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.UseValueTasksCorrectlyMessage_AccessingIncompleteResult), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+            DiagnosticCategory.Reliability,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            s_localizableDescription,
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GeneralDescriptor, UnconsumedDescriptor, DoubleConsumptionDescriptor, AccessingIncompleteResultDescriptor);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // Get the target ValueTask / ValueTask<T> types.  If they don't exist, nothing more to do.
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksValueTask, out var valueTaskType) ||
+                    !compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksGenericValueTask, out var valueTaskOfTType))
+                {
+                    return;
+                }
+
+                // Get the type for System.Diagnostics.Debug.  If we can't find it, that's ok, we just won't use it.
+                compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemDiagnosticsDebug, out var debugType);
+
+                // Process all invocations.  This analyzer works by finding all method invocations that return ValueTasks,
+                // and then analyzing those invocations and their surroundings.
+                compilationContext.RegisterOperationAction(operationContext =>
+                {
+                    // Is this a method invocation?  If not, we're done.
+                    var operation = operationContext.Operation;
+                    var invocation = (IInvocationOperation)operation;
+                    if (invocation.TargetMethod is null ||
+                        invocation.TargetMethod.MethodKind == MethodKind.Constructor) // ignore 'new ValueTask(...)'
+                    {
+                        return;
+                    }
+
+                    // Does the method return ValueTask?  If not, we're done.
+                    if (!valueTaskType.Equals(invocation.TargetMethod.ReturnType) &&
+                        !valueTaskOfTType.Equals(invocation.TargetMethod.ReturnType.OriginalDefinition))
```

Comment:
Do we care about Nullable<ValueTask>?

Reply:
> Do we care about Nullable?

I don't think so.  If you're returning `ValueTask?`, you're off the rails :)  But I hesitate to warn about such a case, as I have no idea why you'd be doing it; might be something we add later.